File: src/services/routeEngine.js
import { connections } from '../data/network.js';

const profileRules = {
  crutch: { label: '목발 이용', transferPenalty: 3, minimumKg: 0, minimumDoor: 0, allowed: station => station.elevator || station.escalator },
  manual: { label: '수동 휠체어', transferPenalty: 6, minimumKg: 0, minimumDoor: 0, allowed: station => station.elevator },
  electric: { label: '전동 휠체어', transferPenalty: 7, minimumKg: 1000, minimumDoor: 90, allowed: station => station.elevator && station.capacityKg >= 1000 && station.doorWidthCm >= 90 }
};

export function validateStation(station, profile) {
  const rule = profileRules[profile];
  return rule.allowed(station);
}

// 인접 리스트: connections(간선 목록)를 역 ID 기준으로 한 번만 인덱싱해 캐시.
// 매번 getRoute()가 호출될 때마다 전체 간선을 선형 스캔하던 O(V×E) 병목 제거.
// connections는 정적 데이터라 배열 참조가 바뀌지 않는 한 재사용 가능.
let adjacencyCache = null;
let adjacencySource = null;

function buildAdjacency(edges) {
  if (adjacencySource === edges && adjacencyCache) return adjacencyCache;
  const map = new Map();
  const addEdge = (from, to, line, minutes) => {
    if (!map.has(from)) map.set(from, []);
    map.get(from).push({ to, line, minutes });
  };
  for (const [a, b, line, minutes] of edges) {
    addEdge(a, b, line, minutes);
    addEdge(b, a, line, minutes);
  }
  adjacencyCache = map;
  adjacencySource = edges;
  return map;
}

// 최소 힙(binary heap): 매 반복마다 큐 전체를 sort()하던 O(V log V) 병목을
// 삽입/추출 각각 O(log V)로 대체.
class MinHeap {
  constructor() {
    this.items = [];
  }

  get size() {
    return this.items.length;
  }

  push(item) {
    const items = this.items;
    items.push(item);
    let i = items.length - 1;
    while (i > 0) {
      const parent = (i - 1) >> 1;
      if (items[parent].cost <= items[i].cost) break;
      [items[parent], items[i]] = [items[i], items[parent]];
      i = parent;
    }
  }

  pop() {
    const items = this.items;
    const top = items[0];
    const last = items.pop();
    if (items.length) {
      items[0] = last;
      let i = 0;
      const n = items.length;
      while (true) {
        const left = i * 2 + 1;
        const right = i * 2 + 2;
        let smallest = i;
        if (left < n && items[left].cost < items[smallest].cost) smallest = left;
        if (right < n && items[right].cost < items[smallest].cost) smallest = right;
        if (smallest === i) break;
        [items[smallest], items[i]] = [items[i], items[smallest]];
        i = smallest;
      }
    }
    return top;
  }
}

// 도착 시 한 번만 역추적하도록 previous 포인터로 경로를 기록.
// (기존에는 간선을 확장할 때마다 path 배열 전체를 복사해 GC 부담이 컸음)
function reconstructPath(previous, key) {
  const path = [];
  let cursor = key;
  while (previous.has(cursor)) {
    const { prevKey, edge } = previous.get(cursor);
    path.push(edge);
    cursor = prevKey;
  }
  return path.reverse();
}

export function getRoute({ from, to, profile, stations, overrides = {} }) {
  const rule = profileRules[profile];
  const stationMap = new Map(stations.map(station => [station.id, { ...station, ...(overrides[station.id] ?? {}) }]));
  const origin = stationMap.get(from);
  const destination = stationMap.get(to);
  if (!origin || !destination) return { error: '출발역과 도착역을 선택하세요.' };
  if (!validateStation(origin, profile)) return { error: `${origin.name}역은 ${rule.label} 기준으로 이용 가능한 편의시설 정보가 없습니다.` };
  if (!validateStation(destination, profile)) return { error: `${destination.name}역은 ${rule.label} 기준으로 이용 가능한 편의시설 정보가 없습니다.` };

  const adjacency = buildAdjacency(connections);
  const startKey = `${from}:`;
  const heap = new MinHeap();
  heap.push({ id: from, cost: 0, line: null, key: startKey });
  const best = new Map([[startKey, 0]]);
  const previous = new Map();

  while (heap.size) {
    const current = heap.pop();
    // best가 이미 더 나은 값으로 갱신된 stale 항목이면 확장하지 않고 건너뜀 (lazy deletion)
    if ((best.get(current.key) ?? Infinity) < current.cost) continue;
    if (current.id === to) return formatRoute(reconstructPath(previous, current.key), stationMap, profile);

    const edges = adjacency.get(current.id) ?? [];
    for (const { to: next, line, minutes } of edges) {
      const nextStation = stationMap.get(next);
      if (!nextStation || !rule.allowed(nextStation)) continue;
      const isTransfer = current.line && current.line !== line;
      const cost = current.cost + minutes + (isTransfer ? rule.transferPenalty : 0);
      const key = `${next}:${line}`;
      if ((best.get(key) ?? Infinity) <= cost) continue;
      best.set(key, cost);
      previous.set(key, { prevKey: current.key, edge: { from: current.id, to: next, line, minutes, isTransfer } });
      heap.push({ id: next, cost, line, key });
    }
  }
  return { error: '현재 조건을 모두 만족하는 경로를 찾지 못했습니다. 다른 출발·도착역 또는 이용 유형을 확인하세요.' };
}

function formatRoute(path, stationMap, profile) {
  const groups = [];
  path.forEach(edge => {
    const last = groups.at(-1);
    if (last?.line === edge.line) last.edges.push(edge);
    else groups.push({ line: edge.line, edges: [edge] });
  });
  const transferStations = path.filter(edge => edge.isTransfer).map(edge => stationMap.get(edge.from));
  const rideMinutes = path.reduce((sum, edge) => sum + edge.minutes, 0);
  const extraMinutes = profileRules[profile].transferPenalty * transferStations.length;
  return {
    groups,
    transferStations,
    rideMinutes,
    totalMinutes: rideMinutes + extraMinutes,
    pathStations: [stationMap.get(path[0].from), ...path.map(edge => stationMap.get(edge.to))],
    profileLabel: profileRules[profile].label
  };
}

File: app/lib/services/route_engine.dart
import '../data/connections.dart';
import '../models/route.dart';
import '../models/station.dart';

/// 전동 휠체어 통과 최소 문 폭(cm). 단, 실측 데이터가 있을 때만 게이팅에 사용.
/// doorWidthCm이 null(실측 전)이면 지어낸 값으로 통과/차단을 결정하지 않는다.
const int kMinDoorCm = 90;

/// 프로필별 이용 가능 판정. (웹 routeEngine.js와 동일 로직)
bool stationAllowed(Station s, MobilityProfile profile) {
  switch (profile) {
    case MobilityProfile.crutch:
      return s.elevator || s.escalator;
    case MobilityProfile.manual:
      return s.elevator;
    case MobilityProfile.electric:
      return s.elevator &&
          s.capacityKg >= 1000 &&
          (s.doorWidthCm == null || s.doorWidthCm! >= kMinDoorCm);
  }
}

/// 다익스트라 경로 탐색 엔진.
/// - 인접 리스트(간선 1회 인덱싱) + 바이너리 최소 힙 + previous 포인터 역추적.
/// - (역, 진입 노선) 쌍을 상태 키로 사용해 환승 페널티를 정확히 반영한다.
class RouteEngine {
  RouteEngine({List<Station>? stations})
      : _stationMap = {for (final s in (stations ?? const [])) s.id: s} {
    _buildAdjacency();
  }

  final Map<String, Station> _stationMap;
  final Map<String, List<_Adj>> _adjacency = {};

  void _buildAdjacency() {
    void add(String from, String to, String line, int min) {
      (_adjacency[from] ??= []).add(_Adj(to, line, min));
    }

    for (final (a, b, line, min) in kConnections) {
      add(a, b, line, min);
      add(b, a, line, min);
    }
  }

  /// 경로를 계산한다. 실패 시 [RouteException]을 던진다.
  RouteResult getRoute({
    required String fromId,
    required String toId,
    required MobilityProfile profile,
  }) {
    final origin = _stationMap[fromId];
    final dest = _stationMap[toId];
    if (origin == null || dest == null) {
      throw RouteException('출발역과 도착역을 선택하세요.');
    }
    if (!stationAllowed(origin, profile)) {
      throw RouteException(
          '${origin.name}역은 ${profile.label} 기준으로 이용 가능한 편의시설 정보가 없습니다.');
    }
    if (!stationAllowed(dest, profile)) {
      throw RouteException(
          '${dest.name}역은 ${profile.label} 기준으로 이용 가능한 편의시설 정보가 없습니다.');
    }

    final startKey = '$fromId:';
    final heap = _MinHeap()..push(_Node(fromId, 0, null, startKey));
    final best = <String, int>{startKey: 0};
    final previous = <String, _Prev>{};

    while (heap.isNotEmpty) {
      final cur = heap.pop();
      if ((best[cur.key] ?? 1 << 30) < cur.cost) continue; // stale
      if (cur.id == toId) {
        return _format(_reconstruct(previous, cur.key), profile);
      }
      for (final adj in _adjacency[cur.id] ?? const <_Adj>[]) {
        final next = _stationMap[adj.to];
        if (next == null || !stationAllowed(next, profile)) continue;
        final isTransfer = cur.line != null && cur.line != adj.line;
        final cost =
            cur.cost + adj.minutes + (isTransfer ? profile.transferPenalty : 0);
        final key = '${adj.to}:${adj.line}';
        if ((best[key] ?? 1 << 30) <= cost) continue;
        best[key] = cost;
        previous[key] = _Prev(
          cur.key,
          RouteEdge(
            from: cur.id,
            to: adj.to,
            line: adj.line,
            minutes: adj.minutes,
            isTransfer: isTransfer,
          ),
        );
        heap.push(_Node(adj.to, cost, adj.line, key));
      }
    }
    throw RouteException(
        '현재 조건을 모두 만족하는 경로를 찾지 못했습니다. 다른 출발·도착역 또는 이용 유형을 확인하세요.');
  }

  List<RouteEdge> _reconstruct(Map<String, _Prev> previous, String key) {
    final path = <RouteEdge>[];
    var cursor = key;
    while (previous.containsKey(cursor)) {
      final p = previous[cursor]!;
      path.add(p.edge);
      cursor = p.prevKey;
    }
    return path.reversed.toList();
  }

  RouteResult _format(List<RouteEdge> path, MobilityProfile profile) {
    final groups = <RouteGroup>[];
    for (final edge in path) {
      if (groups.isNotEmpty && groups.last.line == edge.line) {
        groups.last.edges.add(edge);
      } else {
        groups.add(RouteGroup(edge.line, [edge]));
      }
    }
    final transferStations = [
      for (final e in path)
        if (e.isTransfer) _stationMap[e.from]!
    ];
    final rideMinutes = path.fold<int>(0, (sum, e) => sum + e.minutes);
    final extra = profile.transferPenalty * transferStations.length;
    return RouteResult(
      groups: groups,
      transferStations: transferStations,
      rideMinutes: rideMinutes,
      totalMinutes: rideMinutes + extra,
      pathStations: [
        _stationMap[path.first.from]!,
        for (final e in path) _stationMap[e.to]!,
      ],
      profileLabel: profile.label,
    );
  }

  /// route 결과를 실제 이동 순서로 확장할 노선 구간 리스트로 변환.
  /// isTransfer는 "이 구간의 도착역에서 환승하는가"를 뜻한다 —
  /// 마지막 구간이 아니면(뒤에 탈 노선이 있으면) 도착역은 환승역이다.
  List<RouteLeg> buildLegs(RouteResult result) {
    final groups = result.groups;
    return [
      for (int i = 0; i < groups.length; i++)
        RouteLeg(
          line: groups[i].line,
          fromName: _stationMap[groups[i].edges.first.from]!.name,
          toName: _stationMap[groups[i].edges.last.to]!.name,
          minutes: groups[i].edges.fold<int>(0, (s, e) => s + e.minutes),
          isTransfer: i < groups.length - 1,
        ),
    ];
  }
}

class _Adj {
  final String to;
  final String line;
  final int minutes;
  const _Adj(this.to, this.line, this.minutes);
}

class _Node {
  final String id;
  final int cost;
  final String? line;
  final String key;
  const _Node(this.id, this.cost, this.line, this.key);
}

class _Prev {
  final String prevKey;
  final RouteEdge edge;
  const _Prev(this.prevKey, this.edge);
}

/// 비용 기준 최소 힙.
class _MinHeap {
  final List<_Node> _items = [];
  bool get isNotEmpty => _items.isNotEmpty;

  void push(_Node item) {
    _items.add(item);
    var i = _items.length - 1;
    while (i > 0) {
      final parent = (i - 1) >> 1;
      if (_items[parent].cost <= _items[i].cost) break;
      final tmp = _items[parent];
      _items[parent] = _items[i];
      _items[i] = tmp;
      i = parent;
    }
  }

  _Node pop() {
    final top = _items[0];
    final last = _items.removeLast();
    if (_items.isNotEmpty) {
      _items[0] = last;
      var i = 0;
      final n = _items.length;
      while (true) {
        final left = i * 2 + 1;
        final right = i * 2 + 2;
        var smallest = i;
        if (left < n && _items[left].cost < _items[smallest].cost) smallest = left;
        if (right < n && _items[right].cost < _items[smallest].cost) smallest = right;
        if (smallest == i) break;
        final tmp = _items[smallest];
        _items[smallest] = _items[i];
        _items[i] = tmp;
        i = smallest;
      }
    }
    return top;
  }
}
